src/portable.rs

```diff
@@ -0,0 +1,70 @@
+use std::fs;
+use std::io;
+use std::path::Path;
+
+/// Deletes the contents of `dir_path`, but not the directory iteself.
+///
+/// If `dir_path` is a symlink to a directory, deletes the contents
+/// of that directory.  If `dir_path` does not exist, simply returns
+/// (without deleting anything).
+pub fn remove_dir_contents(dir_path: &Path) -> Result<(), io::Error> {
```

Comment:
I can see why you've written this new function, but I'd like to suggest a different implementation - bearing in mind some future optimisations I have planned. In particular, I want to avoid this serial loop over the read_dir contents at the top layer.

My suggestion is that rather than having allow_nondirectory / directories_only, your new flag to the core workhorse is delete_root; and that the second entry point to it be remove_dir_contents; then the only change needed is to move the remove_item(path, ctx) call from remove_dir_all_recursive to the caller of it, which will allow it to be skipped for the root parent case.

e.g.

```
 ctx.readonly = child.metadata()?.permissions().readonly();
        if child_type.is_dir() {
            remove_dir_all_recursive(&child.path(), ctx)?;
+          remove_item(&child.path(), ctx)
        } else {
            remove_item(&child.path().as_ref(), ctx)?;
        }
    }
    ctx.readonly = dir_readonly;
-    remove_item(path, ctx)
```

And call it from the root case as well.

This of course leaves the posix case needing its own implementation, but I think it will be very shallow at that point.

Reply:
I looked at master, and yes, a different implementation strategy was sensible.  I found it better to split the function rather than using a boolean flag - I prefer to encode things in control flow rather than flag booleans, as that makes analysis (eg type checks) easier.  I hope you like what I have done.